src/lib/pubkey/dilithium/dilithium_common/dilithium_polynomials.h

```diff
@@ -0,0 +1,1609 @@
+         while(ctr < len && pos < buflen)
+            {
+            uint32_t t0 = buf[pos] & 0x0F;
+            uint32_t t1 = buf[pos++] >> 4;
+
+            switch(mode.eta())
+               {
+               case DilithiumEta::Eta2:
+                  {
+                  if(t0 < 15)
+                     {
+                     t0 = t0 - (205 * t0 >> 10) * 5;
+                     a.m_coeffs[offset + ctr++] = 2 - t0;
+                     }
+                  if(t1 < 15 && ctr < len)
+                     {
+                     t1 = t1 - (205 * t1 >> 10) * 5;
+                     a.m_coeffs[offset + ctr++] = 2 - t1;
+                     }
+                  }
+               break;
+               case DilithiumEta::Eta4:
+                  {
+                  if(t0 < 9)
+                     {
+                     a.m_coeffs[offset + ctr++] = 4 - t0;
+                     }
+                  if(t1 < 9 && ctr < len)
+                     {
+                     a.m_coeffs[offset + ctr++] = 4 - t1;
+                     }
+                  }
+               break;
+               }
+            }
+         return ctr;
+         }
+
+      /*************************************************
+      * Name:        fill_poly_uniform_eta
+      *
+      * Description: Sample polynomial with uniformly random coefficients
+      *              in [-ETA,ETA] by performing rejection sampling on the
+      *              output stream from SHAKE256(seed|nonce) or AES256CTR(seed,nonce).
+      *
+      * Arguments:   - Polynomial& a: reference to output polynomial
+      *              - const uint8_t seed[]: byte array with seed of length CRHBYTES
+      *              - uint16_t nonce: 2-byte nonce
+      *              - const DilithiumModeConstants& mode: Mode dependent values.
+      **************************************************/
+      static void fill_poly_uniform_eta(Polynomial& a, const secure_vector<uint8_t>& seed,
+                                        uint16_t nonce, const DilithiumModeConstants& mode)
+         {
+         BOTAN_ASSERT_NOMSG(seed.size() == DilithiumModeConstants::CRHBYTES);
+
+         // TODO: We're deliberately unlocking a secure_vector (to fulfil the
+         //       XOF API). That might be a problem: double-check whether seed
+         //       needs to secure.
+         const std::vector<uint8_t> seed_vec = unlock(seed);
+         auto cipher = mode.XOF_256(seed_vec, nonce);
+
+         secure_vector<uint8_t> buf;
+         buf.resize(mode.poly_uniform_eta_nblocks() * mode.stream256_blockbytes());
+         cipher->cipher1(buf.data(), buf.size());
+         size_t ctr = Polynomial::rej_eta(a, 0, DilithiumModeConstants::N, buf, buf.size(), mode);
+
+         while(ctr < DilithiumModeConstants::N)
+            {
+            std::vector<uint8_t> zero_input;
+            zero_input.resize(mode.stream256_blockbytes());
+            cipher->cipher(zero_input.data(), buf.data(), mode.stream256_blockbytes());
```

Comment:
This could also be `write_keystream`, in which case `zero_input` buffer is not required anymore.